Mathlib/Data/Matroid/Closure.lean

```diff
@@ -519,11 +552,28 @@ lemma basis_iff_indep_closure : M.Basis I X ↔ M.Indep I ∧ X ⊆ M.closure I
   ⟨fun h ↦ ⟨h.indep, h.subset_closure, h.subset⟩, fun h ↦
     (basis_union_iff_indep_closure.mpr ⟨h.1, h.2.1⟩).basis_subset h.2.2 subset_union_right⟩
 
+lemma Indep.inter_basis_closure_iff_subset_closure_inter {X : Set α} (hI : M.Indep I) :
+    M.Basis (X ∩ I) X ↔ X ⊆ M.closure (X ∩ I) :=
+  ⟨Basis.subset_closure,
+    fun h ↦ (hI.inter_left X).basis_of_subset_of_subset_closure inter_subset_left h⟩
+
+lemma Basis.closure_inter_basis_closure (h : M.Basis (X ∩ I) X) (hI : M.Indep I) :
+    M.Basis (M.closure X ∩ I) (M.closure X) := by
+  rw [hI.inter_basis_closure_iff_subset_closure_inter] at h ⊢
+  exact (M.closure_subset_closure_of_subset_closure h).trans (M.closure_subset_closure
+    (inter_subset_inter_left _ (h.trans (M.closure_subset_closure inter_subset_left))))
+
 lemma Basis.eq_of_closure_subset (hI : M.Basis I X) (hJI : J ⊆ I) (hJ : X ⊆ M.closure J) :
     J = I := by
   rw [← hI.indep.closure_inter_eq_self_of_subset hJI, inter_eq_self_of_subset_right]
   exact hI.subset.trans hJ
 
+lemma Basis.insert_basis_insert_of_not_mem_closure (hIX : M.Basis I X) (heI : e ∉ M.closure I)
+    (heE : e ∈ M.E := by aesop_mat) : M.Basis (insert e I) (insert e X) := by
+  refine hIX.insert_basis_insert ?_
```

Comment:
Does `apply` work?

Reply:
Presumably yes. But I always think of `apply` as doing more work than `refine` and being more opaque, so I prefer the latter.